src/prices/ValueInterpreter.sol

```diff
@@ -0,0 +1,155 @@
+    /// either using live or canonical conversion rates
+    function __calcAssetValue(
+        address _baseAsset,
+        uint256 _amount,
+        address _quoteAsset,
+        bool _useLiveRate
+    )
+        private
+        view
+        returns (uint256 value_, bool isValid_)
+    {
+        IRegistry registry = IRegistry(REGISTRY);
+
+        // Check if registered _asset first
+        // TODO: Consider checking against IPriceSource instead of Registry
+        if (registry.assetIsRegistered(_baseAsset)) {
+            return __calcPrimitiveValue(_baseAsset, _amount, _quoteAsset, _useLiveRate);
+        }
+
+        // Else use derivative oracle to get value via underlying assets
+        else if (registry.derivativeToPriceSource(_baseAsset) != address(0)) {
+            return __calcDerivativeValue(_baseAsset, _amount, _quoteAsset, _useLiveRate);
+        }
+
+        // If not in Registry as an asset or derivative
+        else return (0, false);
+    }
+
+    /// @dev Helper to calculate the value of a primitive (an asset that has a price
+    /// in the primary pricefeed) in an arbitrary asset.
+    function __calcPrimitiveValue(
+        address _primitive,
+        uint256 _amount,
+        address _quoteAsset,
+        bool _useLiveRate
+    )
+        private
+        view
+        returns (uint256 value_, bool isValid_)
+    {
+        IPriceSource priceSource = IPriceSource(IRegistry(REGISTRY).priceSource());
+
+        uint256 rate;
+        if (_useLiveRate) {
+            (rate, isValid_) = priceSource.getLiveRate(_primitive, _quoteAsset);
+        }
+        else {
+            (rate, isValid_,) = priceSource.getCanonicalRate(_primitive, _quoteAsset);
+        }
+
+        value_ = __calcConversionAmount(_primitive, _amount, rate);
+    }
+
+    /// @dev Helper to calculate the value of a derivative in an arbitrary asset.
+    /// Handles multiple underlying assets (e.g., Uniswap and Balancer pool tokens).
+    /// Handles underlying assets that are also derivatives (e.g., a cDAI-ETH LP)
+    function __calcDerivativeValue(
+        address _derivative,
+        uint256 _amount,
+        address _quoteAsset,
+        bool _useLiveRate
+    )
+        private
+        view
+        returns (uint256 value_, bool isValid_)
+    {
+        address derivativePriceSource = IRegistry(REGISTRY).derivativeToPriceSource(_derivative);
+        (
+            address[] memory underlyings,
+            uint256[] memory rates
+        ) = IDerivativePriceSource(derivativePriceSource).getRatesToUnderlyings(_derivative);
+
+        for (uint256 i = 0; i < underlyings.length; i++) {
+            uint256 underlyingAmount = __calcConversionAmount(underlyings[i], _amount, rates[i]);
+            (
+                uint256 underlyingValue,
+                bool underlyingIsValid
+            ) = __calcAssetValue(underlyings[i], underlyingAmount, _quoteAsset, _useLiveRate);
+
+            if (isValid_ && !underlyingIsValid) isValid_ = false;
```

Comment:
can the conditional be shortened to just `if (!underlyingIsValid)`? Also, is it possible for `isValid_` to be returned as `true`, even with the change I mentioned?

Reply:
Ah yes, I forgot to assign `isValid_` to `true` before the loop. With L142, I was trying to not re-assign to `false` redundantly if there are multiple invalid rates, but it's clearer with your suggestion, so I've done that.

Fixed below.